### PR TITLE
[BT-5168-event-feeds] Implement Event Feeds for JVM

### PR DESCRIPTION
**Pull request by findgriffin**
### Description

This change implements the new Event Feeds feature.

### Motivation and context

Event Feeds needs support in all drivers to be useful to customers.

### How was the change tested?

Unit and E2E tests, I eyeballed the coverage and I think most of the new code is covered.

### Change types

* - [ ] Bug fix (non-breaking change that fixes an issue)
* - [x] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [x] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (pnwpedro) [inline src/main/java/com/fauna/client/FaunaClient.java:531]: if i use this API to get an iterator, it's not clear to me how i'd track the cursor and continue my feed where i left off. in python there's a FeedPage that's returned that allows for this. i think with this API it ends up being a pagination until the end of the feed, but then you cannot continue where you left off on the next run (unless i'm missing something?)

**Review by pnwpedro — CHANGES_REQUESTED** [review]
A solid start, but a few big questions about the impl.

File: README.md
Comment (jrodewig) [reply README.md]: Thanks. I've simplified this with [`b18b13e` (#154)](https://github.com/fauna/fauna-jvm/pull/154/commits/b18b13e7414cec9f8d74398040493a8f350b85c1). Good feedback!

File: src/main/java/com/fauna/feed/FeedRequest.java
Comment (findgriffin) [reply src/main/java/com/fauna/feed/FeedRequest.java:8]: Yeah, we could do that, it looks like `FeedRequest` and `StreamRequest` are the same, except that `FeedRequest` has `pageSize` we could pass `pageSize` in separately, or just make it optional?

File: src/main/java/com/fauna/feed/FeedSuccess.java
Comment (findgriffin) [reply src/main/java/com/fauna/feed/FeedSuccess.java:28]: Yeah, also we don't have the concept of a `FeedResponse` that we subclass into success/failure, like for query, and I'm not sure that the feed (or stream) spec defines what a failure should look like. Should I re-use `QueryFailure`? Should we rename `QueryFailure` to `FaunaFailure` should we have classes `[Query|Stream|Feed]Failure` that subclass `FaunaFailure`?

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (findgriffin) [reply src/main/java/com/fauna/client/FaunaClient.java:531]: I guess I was thinking that if you want to continue waiting for new events, that's not a feed, that's a stream. I think you can get the latest TS off the feed events and restart that way but maybe this use case is something to flesh out in the E2E tests.

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (findgriffin) [reply src/main/java/com/fauna/client/FaunaClient.java:113]: I think I wasn't sure if we get the latest timestamp back like we do for a query.

File: src/main/java/com/fauna/feed/FeedRequest.java
Comment (findgriffin) [reply src/main/java/com/fauna/feed/FeedRequest.java:8]: This is what the JS PR currently has:

```
export class StreamToken {
export interface EventSource {
  readonly token: string;
}

export function isEventSource(value: any): value is EventSource {
  if (typeof value.token === "string") {
    return true;
  }

  return false;
}

export class StreamToken implements EventSource {
  readonly token: string;

  constructor(token: string) {
    this.token = token;
  }
}

/**
 * A class to represent a page of events from a Fauna stream.
 */
export class FeedPage<T extends QueryValue> {
```

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (pnwpedro) [reply src/main/java/com/fauna/client/FaunaClient.java:113]: I think the core team will have a definitive answer here, but probably checking the python driver is a good call since @erickpintor implemented that

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (pnwpedro) [reply src/main/java/com/fauna/client/FaunaClient.java:531]: well, yeah, but it's polling instead of streaming. so if i'm running a lambda and i want to poll a feed for the events since the last event i saw, i need a way of persisting data about the last event i saw (whether that's a cursor or a start ts).

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (erickpintor) [reply src/main/java/com/fauna/client/FaunaClient.java:113]: We're not updating the client's last seen txn time based on streaming and event feeds at the moment.

File: src/main/java/com/fauna/client/FeedIterator.java
Comment (erickpintor) [inline src/main/java/com/fauna/client/FeedIterator.java:61]: If I understand this code correctly we're returning the result of the already requested page while issuing the next page request in the background. If so, I think that's a bit counter intuitive to customers since they lose control on rather to call the next page or not.

File: src/main/java/com/fauna/feed/FeedRequest.java
Comment (erickpintor) [inline src/main/java/com/fauna/feed/FeedRequest.java:18]: We usually accept either cursor or start time, not both.

File: README.md
Comment (erickpintor) [inline README.md:545]: Q: could we have a more idiomatic overload here using `java.time` instead of a long?

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (pnwpedro) [reply src/main/java/com/fauna/client/FaunaClient.java:531]: I think we're actually okay here. I was looking at this and for some reason missing that the iterator returns pages (FeedSuccess) or on flatten() events. Both of these contain the cursor.

File: src/main/java/com/fauna/feed/FeedSuccess.java
Comment (pnwpedro) [reply src/main/java/com/fauna/feed/FeedSuccess.java:28]: > Is this the same concept as FeedPage in the python driver?

looks like it is. i'll leave it up to you where you want to take this. you can leave it as FeedSuccess if that's natural to you

File: src/main/java/com/fauna/response/StreamEvent.java
Comment (findgriffin) [reply src/main/java/com/fauna/response/StreamEvent.java:23]: I'm going to put up a renaming PR, and I'll address this there.

File: src/main/java/com/fauna/client/FeedIterator.java
Comment (findgriffin) [reply src/main/java/com/fauna/client/FeedIterator.java]: Eric and I chatted about this, I added `E2EFeedsTest.manualFeed` to explicitly show how a user can control the requests being sent to Fauna, but I left the implementation of `FeedIterator` as-is.

File: README.md
Comment (findgriffin) [reply README.md:545]: This got moved over to `FeedOptions` and morphed into a `Duration` in the process.

File: src/main/java/com/fauna/feed/FeedSuccess.java
Comment (findgriffin) [reply src/main/java/com/fauna/feed/FeedSuccess.java]: I renamed it to FeedPage.

File: src/main/java/com/fauna/feed/FeedPage.java
Comment (pnwpedro) [inline src/main/java/com/fauna/feed/FeedPage.java:25]: I think this class and FeedRequest might deserve some unit tests because of the token logic, but I'm fine if that comes later.

**Review by pnwpedro — APPROVED** [review]
🥳  lgtm!

File: src/main/java/com/fauna/feed/FeedPage.java
Comment (findgriffin) [reply src/main/java/com/fauna/feed/FeedPage.java:25]: Agree.

File: src/main/java/com/fauna/client/FaunaClient.java
Comment (findgriffin) [reply src/main/java/com/fauna/client/FaunaClient.java:133]: Will do.

**Pull request merged**
